x-pack/plugins/lens/public/trigger_actions/open_lens_config/create_action_helpers.ts

```diff
@@ -31,20 +32,41 @@ export const [getDatasourceMap, setDatasourceMap] = createGetterSetter<
   Record<string, Datasource<unknown, unknown>>
 >('DatasourceMap', false);
 
-export function isCreateActionCompatible(core: CoreStart) {
-  return core.uiSettings.get(ENABLE_ESQL);
+export async function isCreateActionCompatible(
+  core: CoreStart,
+  editorFrameService: EditorFrameService
+) {
+  let isCompatible: boolean = core.uiSettings.get(ENABLE_ESQL);
+
+  if (isCompatible && (!getDatasourceMap() || !getVisualizationMap())) {
+    const [visualizationMap, datasourceMap] = await Promise.all([
+      editorFrameService.loadVisualizations(),
+      editorFrameService.loadDatasources(),
+    ]);
+
+    if (visualizationMap && datasourceMap) {
+      setDatasourceMap(datasourceMap);
+      setVisualizationMap(visualizationMap);
+    } else {
+      isCompatible = false;
+    }
+  }
+
+  return isCompatible;
```

Comment:
Why introduce a side effect in the compatibility check?
Would it be best to move it within the execution flow?

Reply:
After speaking with @dej611 on slack we decided it's acceptable to have this side effect in the execution flow and halt the execution flow if it so happens that the required data dependency isn't available. https://github.com/elastic/kibana/pull/187797/commits/1017c768fe5bac38b1b65514c3300546a6c8ed6c addresses this change.